patching-explainer.md

```diff
@@ -0,0 +1,221 @@
+# Patching (Partial, interleaved HTML streaming)
+
+## Motivation
+Streaming of HTML existed from the early days of the web, serving an important purpose for perceieved performance when loading long articles etc.
+However, it always had the following major constraints:
+1. HTML content is streamed in DOM order.
+2. After the initial parsing of the document, streaming is no longer enabled.
+
+Use cases like updating an article without refreshing the page, or for streaming the contents of a scrollable container after it is loaded,
+are accomplished today by custom JS that uses the DOM APIs, or by JS frameworks that abstract these away.
+
+For example, React streams content out of order by injecting inline `<script>` tags that modify the already parsed DOM.
+
+This proposal introduces partial out-of-order HTML streaming as part of the web platform.
+
+## Patching
+A "patch" is a stream of HTML content, that can be injected into an existing position in the DOM.
+A patch can be streamed directly into that position using JavaScript, and multiple patches can be interleaved in an HTML document, allowing for out-of-order content as part of an ordinary HTML response.
+
+## Anatomy of a patch
+
+A patch is a [stream](https://developer.mozilla.org/en-US/docs/Web/API/Streams_API) that targets a [parent node](https://developer.mozilla.org/en-US/docs/Web/API/Node) (usually an element, but potentially a shadow root).
+It can handle strings or bytes. When it receives bytes, it decodes them using the node's document's charset. Anything other than strings or bytes is stringified.
+
+When a patch is active, it is essentially a [WritableStream](https://developer.mozilla.org/en-US/docs/Web/API/WritableStream) that feeds a [fragment-mode parser](https://html.spec.whatwg.org/multipage/parsing.html#html-fragment-parsing-algorithm) with strings from that stream.
+It is similar to calling `document.write()`, scoped to an node.
+
+## One-off patching
+
+The most atomic form of patching is opening a container node for writing, creating a `WritableStream` for it.
+This can be done with an API as such:
+```js
+const writable = elementOrShadowRoot.patch();
+byteOrTextStream.pipeTo(writable);
+```
+
+A few details about one-off patching:
+- Trying to patch an element that is currently being patched would abort the original stream.
+- It is unclear whether patching should execute `<script>` elements found in the patch. It would potentially be an opt-in. See #40.
```

Comment:
Seems like this won't be linked correctly unless the link is spelled out:

```suggestion
- It is unclear whether patching should execute `<script>` elements found in the patch. It would potentially be an opt-in. See https://github.com/WICG/declarative-partial-updates/issues/40.
```